### PR TITLE
Add check of graph output in Bert Fusions

### PR DESCRIPTION
**Pull request by tianleiwu**
**Description**: 

Node.GetOutputEdges() does not include graph output. Additional check is added for each internal node (in a subgraph to be fused): Do not fuse if internal node has graph output.

**Motivation and Context**
- Why is this change required? What problem does it solve?
- If it fixes an open issue, please link to the issue here.

Address the bug in issues https://github.com/microsoft/onnxruntime/issues/4110

### CODE REVIEW

File: onnxruntime/core/optimizer/utils.h [hunk 1]
```diff
@@ -66,5 +66,10 @@ int32_t IndexOfNodeInput(const Node& node, const NodeArg& node_arg);
 */
 bool IsSupportedDataType(const Node& node, const std::vector<std::string>& supported_data_types);
 
+/** Check whether node's output edges count is expected.
+@remarks graph output is not included in output edges, and by defult graph output is not allowed.
+*/
+bool CheckOutputEdges(const Graph& graph, const Node& node, size_t expected_output_edges, bool no_graph_output = true);
+
```
Comment (skottmckay) [inline onnxruntime/core/optimizer/utils.h:73]: Nothing passes false for no_graph_output so why does that need to be a parameter?

Comment (askhade) [reply onnxruntime/core/optimizer/utils.h:73]: at the very least change the name to make it clear... something like not_in_graph_outputs or not_a_graph_output

File: onnxruntime/core/optimizer/attention_fusion.cc [hunk 2]
```diff
@@ -413,16 +413,14 @@ bool AttentionFusion::FuseSubGraph(Node& layer_norm, const Node& add_after_layer
...
-      matmul.GetOutputEdgesCount() != 1 ||
-      reshape.GetOutputEdgesCount() != 1 ||
-      transpose.GetOutputEdgesCount() != 1 ||
-      qkv_matmul.GetOutputEdgesCount() != 1 ||
-      v_transpose.GetOutputEdgesCount() != 1 ||
-      v_reshape.GetOutputEdgesCount() != 1 ||
-      v_add.GetOutputEdgesCount() != 1 ||
-      v_matmul.GetOutputEdgesCount() != 1 ||
-      v_root.GetOutputEdgesCount() != 4) {
+  // Internal nodes of attention subgraph only allow edges within the subgraph, and no graph output is allowed.
```
Comment (askhade) [inline onnxruntime/core/optimizer/attention_fusion.cc:416]: should these nodes be exempted from the number of output edges check as well? .i.e. add.GetOutputEdgesCount() != 1

File: onnxruntime/core/optimizer/utils.h (hunk 1 above)
Comment (tianleiwu) [reply onnxruntime/core/optimizer/utils.h:73]: Remove the parameter and update comments.

File: onnxruntime/core/optimizer/attention_fusion.cc (hunk 2 above)
Comment (tianleiwu) [reply onnxruntime/core/optimizer/attention_fusion.cc:416]: There are comments that the constraints are removed for outside nodes (add, matmul and v_root) not to be fused.

**Pull request merged**